### ISSUE TITLE
[Suggestion] Callbacks in Trainer

### ISSUE THREAD

**Issue by keitakurita** [allenai/allennlp#2587]
Hello,

First off, thank you so much for creating and maintaining this amazing library! I'm really loving allennlp so far, but there's one aspect that I think could use some improvements: the Trainer. I think the adoption of Callbacks would make the code much more readable, maintainable, and extendable.

**Is your feature request related to a problem? Please describe.**
- The Trainer code feels bloated and is hard to navigate due to the sheer amount of bookkeeping that is taking place regarding tensorboard, checkpointing, etc. 
- Adding extra behavior to the trainer (e.g. custom logging, adding semi-supervised steps, performing custom sanity checks during training) requires the modification of the training code. It would be better if the user could simply inject this behavior using an external class

**Describe the solution you'd like**
Frameworks like keras and fast.ai have adopted the Callback pattern to deal with this problem. In fact, checkpointing and tensorboard were both callbacks in keras. Refactoring the trainer to adopt the callback system for a lot of the bookkeeping would make custom training behavior easier to implement.

**Describe alternatives you've considered**
I can't think of any better alternatives at the moment: the fact that both keras and fast.ai which put heavy emphasis on the end user experience adopt this pattern seem to be a strong indication of its merits. That being said, I would imagine the allennlp team would have considered callbacks at some time, so if there is a reason you are avoiding this pattern, I would love to know why!

If Callbacks seem like a good addition, I'd love to file a PR on this issue, but since this would be a major refactoring I won't be able to finish it any time soon. Thanks!

**Comment by matt-gardner** [comment]
Related to this, I thought that the decorator design pattern used by [pytorch/ignite](https://github.com/pytorch/ignite) looked appealing.  It's similar to callbacks, but makes use of decorators in a way that I think is nice.  It might be hard to get that to work with our config-driven experiments, though...  I think this general idea is good, but a lot of work, as you say @keitakurita.  @joelgrus, what do you think?  I think the refactoring you recently did should make doing this further refactoring a lot easier than it would have been.

So, in my opinion, this would be a great addition, but not one that is at the top of our priority list right now.  If you or anyone else wants to take this on and contribute a PR (or set of PRs that incrementally get us there), that'd be great, though it'd be good to hear positive feedback from at least one other allennlp dev before starting on this.  It'd also be good to agree on a general approach for how to do this (explicitly-passed callback functions, decorator functions, etc.).

**Comment by matt-gardner** [comment]
E.g., are callbacks something you register?  Or are they a method you override / extend the `Trainer` for?  There are a number of different ways to do this, and I'm not sure what the right way is without thinking it through a whole lot more.

**Comment by keitakurita** [comment]
Thanks for the prompt reply! The ignite library seems interesting, I'll take a look at it.
My current line of thinking is to make Callbacks explicitly passed objects like in the case of Keras and fast.ai. The `Callback` class could be a `Registrable` and you could pass a list of callback configurations when instantiating the `Trainer`

e.g. 
```
"trainer": {
    "optimizer": {
        "type": "adam",
        "lr": 0.001
    },
    "validation_metric": "+f1-measure-overall",
    "callbacks": [
          {"type": "tensorboard", "serialization_dir": "data",
            "should_log_parameter_statistics": true}
     ]
```

This seems like a much easier extension of the current way the library is implemented (the Tensorboard and Checkpointer are already their own classes, so moving them to a separate callback class seems much more trivial). It also allows users to configure and construct their own callbacks in an allennnlp-like way.
This could break compatibility with previous configurations though, so we'll need to be careful to enable users to specify tensorboard/checkpointing like before or gradually phase out with deprecation warnings.

With regards to the decorator approach, we could probably extend the object-based callback system to use decorators as syntactic sugar (e.g. `@trainer.on(Events. ITERATION_COMPLETED)` appends a callback object to the trainer that executes the function when each iteration is completed). It would seem to me like keeping callbacks as a `Registrable` as the core implementation mechanism is the most sensible approach. 

What do you think?

**Comment by WrRan** [comment]
I can't agree with this suggestion anymore. I am also considering this issue recently, and I prefer to call this method an event-based Trainer.

I think that the current Trainer has taken on too many responsibilities, which brings us a lot of trouble:
1. Bring some difficulties to maintenance.
2. Customizing your own Trainer can only be achieved through inheritance, which not only means a lot of template code, but also makes implementation difficult.
3. Some of the functions in the Trainer that should be abstracted are not abstracted separately, resulting in some repetitive, highly similar code. For example, the function to load data into multiple GPUs should be abstracted for use by other commands, such as `evaluate`, `predict` and `elmo`. (#2578)
4. `Trainer` does not expose enough parameters for us to configure. We can't directly use the configuration files to finely configure the behavior of `Trainer`, such as, customize the format of the log through the configuration file. This is contrary to the design goal of AllenNLP, a configuration-oriented library. (This is how I feel about using AllenNLP. And I hope "configuration-oriented" reflects the design goals of AllenNLP.)

Fortunately, I think the **event-based Trainer** can solve most of the above problems, and it can also solve other issues, such as #1536. But the downside of event-based Trainer is that implementation may be not easy. 

As of now, I have some points to implement that may need to be considered (*may not be complete or even wrong*).

As for (pytorch/ignite)[https://github.com/pytorch/ignite], it did give me some inspiration, too. But I think its use of decorators, aiming to conveniently set the callback function, may be not appliable to our "configuration-oriented" goal.

In order to implement an **event-based** `Trainer`, we may need
1. Summarize and organize the function points involved in Trainer;
2. Design the base class `Event` (inherited from `Registable`) according to share inputs/outputs of all function points while adhering to the event-based design principles;
3. Design a more detailed subclass for each function based on the status, target, parameters, etc. of the function;
4. Re-determine the responsibilities of the base class `Trainer` based on the share process of all relevant commands (such as `evaluate`, `predict` and `dry-run`) while adhering to the event-based design principles;
5. Design and implement the base class `Trainer`, the concrete sub-class used by default;
6. Provide subclasses with specific event classes to `evaluate`, `provide` and `dry-run`.

Since I have less experience in developing event-based programs, it is uncertain whether the above methods are feasible or not.
But there is one thing can be expected: it will take a lot of work to get the job done.

What do you think?

**Comment by keitakurita** [comment]
@WrRan Thanks for the suggestion! Sorry if I misunderstand your point, but here's my take:

In the callback mechanism, the `Trainer` calls callbacks when certain "events" (e.g. `batch_begin`, `epoch_end`) take place, requesting that each individual callback object executes whatever action is required at that moment (e.g. logging histogram statistics). The events and subsequent processing are triggered by the `Trainer` and an internal `CallbackManager`.

I think having a generic `Event` class would be an extension of the callback mechanism. Instead of having a fixed set of endpoints for executing callbacks such as when the iteration begins, the trainer emits events at various timings (that can be modified by the user) which result in the callbacks being called on the event. For instance, a callback that runs at the beginning of each epoch could check the type of the event and only execute if the event is of type `EPOCH_BEGIN`. Passing `Event` objects would allow the `Trainer` to pass important information like the loss values cleanly as well.

I think creating a single object for each event that handles all the processing could make the codebase messier. With the `Callback` mechanism, we would only need to implement one `TensorboardCallback` for handling all tensorboard-related functionality. In contrast, if we had to implement one class for each event, we would have to change multiple classes whenever we wanted to add changes to the tensorboard functionality.

**Comment by WrRan** [comment]
Sorry about the unclear comment.

The class `Event` in my discussion is not responsible for *handling specific logic*. Its **role** is to identify the occurrence of certain events and carry some relevant information.

As for who is responsible for *handling certain events*, we should indeed use a `Callback` to handle it. (I prefer to call it  `Handler`.)

Thus our views should be consistent. What do you think?

Personally, I might prefer that the final configuration file is as follows.
```json
{
    "trainer": {
        "num_epochs": 30,
        "patience": 10,
        "cuda_device": 0,
        "events": {
            "iteration_begin": {
                "0": [{"type": "metric_reporter"}],
                "1": [{"type": "tensorboard_writer"}]
            },
            "epoch_end": {
                "0": [{"type": "metric_reporter"}]
                // ...
            },
        }
        // ...
    }
}
```

In the above configuration file, the value of the key ·events` is used to specify the `Handler`s used to process each event. A key such as `iteration_begin` specifies the class of event. A key such as `"0"`, `"1"` specifies the priority of the `Handler`s.

**Comment by joelgrus** [comment]
I think it makes a lot of sense to do something like this (even after the refactor the trainer is way too complicated), but I think it's something where we have to really invest in some up-front design work before we just start coding. It's possible this work could make sense as a Q2 goal, let me think about it some more.

**Comment by WrRan** [comment]
I am willing, even a little excited, to participate in this restructuring, with your permission, of course. :)

**Comment by keitakurita** [comment]
@WrRan I see, thanks for the explanation! That's an interesting idea, but here are some problems that I think might arise:
- It involves a lot of boilerplate: you already need to write "metric_reporter" twice in the config file. I would want to just specify that I want to report metrics without having to write every timing it runs at
-The user would need to know the appropriate timing to call each Handler
-The Handler would still need to also know what to do at each timing (e.g. we want to record different metrics at iteration begin and epoch end). 
- Coordination between processing at different timings becomes difficult (e.g. saving weights before the optimization step, then using those weights to compute statistics after the optimization step)

I think configuration files are a good place for declarative programming and expressing high-level intentions. The complex processing of `Callbacks`/`Handlers` seems very procedural, and I would personally prefer to defer that complexity to Python code in the form and configure them as `Registrables`. I understand there are other opinions though, and would love to discuss what the best design for this feature is.

@joelgrus I would also love to participate in this and contribute something to this wonderful library!

**Comment by WrRan** [comment]
@keitakurita Thanks for your feedback. 
 1. The config of the example in https://github.com/allenai/allennlp/issues/2587#issuecomment-471600270  is just to show the format of the final configuration file I expected. Since the functions involved in the training processing are really many, we should reasonably design to avoid boilerplate. Thus, I mentioned earlier in https://github.com/allenai/allennlp/issues/2587#issuecomment-471323518 that we need to provide some specific subclasses for commands `train`, `evaluate`, `predict` and `elmo`. In other words, we should provide a subclass of `Trainer` for each command (or use case). The subclass should provide common `Handler`s to cover frequently-used events. In my opinion, these commands are all `Trainer`s. The difference between them is that the `Handler` called when each `Event` occurs is different.

2. Yes. Making event-based `Trainer` work is not a simple task.

---
For the first issue, in order to more clearly state how to avoid potential boilerplate, I list some configs under use cases that may be encountered as follows.
 - How to **add** some `Handler`s for *a certain priority* under *a specific event*:
    This use case includes two more granular user scenarios: add _before_ default `Handler`s or add _after_ default `Handler`s.
```json
{
    "trainer": {
        "events": {
            "iteration_begin": {
                "+0": [{"type": "metric_reporter"}],
                "1+": [{"type": "tensorboard_writer"}]
            }
        }
    }
}
```
In the code snippet above, key `"+0"` indicates adding some `Handler`s _before_ default ones for the priority of `"0"` under event `iteration_begin` and key `"1+"` indicates adding some `Handler`s _after_ default ones for the priority of `"1"` under event `iteration_begin`.
 - How to **delete** some `Handler`s for *a certain priority* under *a specific event*:
```json
{
    "trainer": {
        "events": {
            "iteration_begin": {
                "-0": [{"type": "metric_reporter"}]
            }
        }
    }
}
```
In the code snippet above, key `"-0"` indicates deleteing the `Handler` whose type is `metric_reporter` for the priority of `"0"` under event `iteration_begin`.
Additionally, follow Tim Peters's design philosophy: There should be one-- and preferably only one --obvious way to do it. We do not support or complain to users about the configuration taking the form of `"0-": [{"type": "metric_reporter"}]`.
 - How to **replace** some `Handler`s for *a certain priority* under *a specific event*:
```json
{
    "trainer": {
        "events": {
            "iteration_begin": {
                "^0": {
                    "metric_reporter": {"type": "customized_metric_reporter"},
                    "tensorboard_writer": {"type": "customized_tensorboard_writer"}
                }
            }
        }
    }
}
```
Similarly, we can replace some specific `Handler`s by using the symbol `^`. Unlike `add` or `delete`, we don't have to think about the order of the `Handler`s when we replace them, so we use _dict_ instead of _list_.
 - How to **delete** or **replace** some `Handler`s for *all priorities* under *a specific event*:
```json
{
    "trainer": {
        "events": {
            "iteration_begin": {
                "-": [{"type": "tensorboard_writer"}],
                "^": {
                    "metric_reporter": {"type": "customized_metric_reporter"}
                }
            }
        }
    }
}
```
As shown in the code snippet above, we delete or replace some `Handler`s for all priorities under a specific event respectively by using `"-"` or `"^"` without the priority symbol.
 - How to **delete** or **replace** some `Handler`s under *all events*:
```json
{
    "trainer": {
        "events": {
            "-": [{"type": "tensorboard_writer"}],
            "^": {
                "metric_reporter": {"type": "customized_metric_reporter"}
            }
        }
    }
}
```
Similarly, we remove or replace some `Handler`s under all events by specifying the value of the key `"-"` or `"^"` in the JSON object `events`.
 - (not support) How to *add* some `Handler`s for *all priorities* under *a specific event*.
 - (not support) How to *add* some `Handler`s under *all events*.
The latter two requirements are not clear enough, so I tend not to support them. For example, when a certain `Handler` of the same type already exists in a priority under an event, this requirement does not clearly indicate what to do.

---
Another reason not using decorators to conveniently set the `Handler`s.

Although due to the use of decorators, we can easily specify the corresponding event when defining the method. But it is precisely because of this practice that we can't see all the Handlers that handle all events in a centralized code block.

This means that we believe that **ease of development is more important than ease of understanding**, and I do **not** think this is a good idea. This is also one reason why I use `allennlp` instead of `ignite`.

To more intuitively illustrate the shortcomings of using a decorator, imagine a similar scenario below. New users want to quickly know if `allennlp` has provided a bi-attention mechanism without browsing the entire codebase of `allennlp`.

What do you think?

**Comment by joelgrus** [comment]
I am worried that this is all way more complicated than what I want. my goal is for the trainer to be even simpler, and so I'm not inclined to create a cryptic custom configuration language for it.

I do like the idea of registrable callbacks, but again I would like to keep them simple.

**Comment by matt-gardner** [comment]
I'm with @joelgrus here. This should simplify the code and make it easier to extend, not add a bunch of complex configuration. I think the first step is identifying which pieces in the existing trainer should be moved to callbacks, and how exactly they should be called. Once that is done, we can make the change, and make those callbacks registrable, allowing others to add their own if they want.

**Comment by matt-gardner** [comment]
Also, having thought about it a bit more, I agree with the consensus here that event handling with decorators is incompatible with how we configure things. It might be cute for when you write your own script instead of using our config stuff, but it definitely shouldn't be a major design consideration as we think about an initial implementation.

**Comment by sudarshan85** [comment]
Hi,
I'm curious whether it might be possible to just use ignite under the hood. This might be totally stupid comment as I don't know the internals of allennlp or ignite. But I like both the libraries, so I just wanted to put this out here.

**Comment by epwalsh** [comment]
Just saw this Twitter thread about callbacks today and thought it would be helpful for this discussion: https://mobile.twitter.com/math_rachel/status/1105528235687727104

**Comment by joelgrus** [comment]
thanks, that's mostly along the lines of what I was thinking

**Comment by keitakurita** [comment]
@matt-gardner @joelgrus 
The fast.ai callback system is what I was envisioning as well, except I think it might be better to allow users to define their own events as well (though it would not be something the allennlp library would have to support). We could just have the callback implement an "on event" method that dispatches the method.

In my opinion, given the Tensorboard and Checkpointer are both already their own classes, it seems most natural to refactor these out as callbacks. This would reduce the code within the Trainer class substantially and would also be a good example for making design decisions. What do you think?

**Comment by matt-gardner** [comment]
I'm not sure how we can have a system where a user defines their own events - by definition, the events have to be triggered in the main `Trainer` code, and there isn't a way for a user to add events there without overriding the `Trainer` (which, to be fair, they can).

As far as exactly where to start and what to do, I'll defer to @joelgrus.

**Comment by WrRan** [comment]
Maybe I underestimated the complexity of developing those features. In my opinion, the most difficult part of this refactoring is to determine the respective responsibilities of `Event`, `Handler` (or `Callback`) and `Trainer` (i.e. [`engine`](https://github.com/pytorch/ignite/blob/master/ignite/engine/engine.py#L34) in [pytorch/ignite](https://github.com/pytorch/ignite)), and coordinate their functions.

But I do agree that we should give priority to the callback mechanism now, because _premature optimization is the root of all evil (or at least most of it) in programming_, and we still don't know whether the above use cases are frequent and core needs.

As for how to let users customize the event, maybe [this PR in `ignite`](https://github.com/pytorch/ignite/pull/220) can provide some help.

**Comment by joelgrus** [comment]
so I still haven't thought about this a ton, but my initial instinct would look something like the following:

1. We have some notion of `Callback(Registrable)`
2. there is some kind of `CallbackHandler`, which associates callbacks with "events"
3. the training loop has some default set of events (e.g. "begin epoch", "before forward", etc...) and it sends a message (but with what payload?) to the callback handler on each of these events
4. we move all of the functionality around "log to tensorboard" and "checkpointing" and stuff like that into callbacks so that the trainer itself is clean and minimal
5. we figure out a simple way to configure more / fewer callbacks using the config file

one thing that leaps out at me as potentially tricky is that each callback will likely have specific data it needs, and it might be a pain to handle that safely and robustly.

**Comment by matt-gardner** [comment]
I was imagining each event type having its own API and handler, with a specific set of things passed to the handler for that event.

**Comment by keitakurita** [comment]
I personally think that it might be best to just pass a generic dictionary of data to each handler on each event as the payload. The default `Trainer` would fill the dictionary with basic information that is necessary to support the `Callbacks` we support by default (e.g. the loss value of the current epoch, the best loss value so far, etc.), but users could also subclass the `Trainer` and have it pass additional information to support custom `Callbacks`. Specifying the API of each handler in advance seems difficult and slightly contrary to the goal of a flexible training loop. We might want a different API just for callbacks that interact with the loss though, since users might want to modify the loss values during training via a callback.

**Comment by WrRan** [comment]
As far as I am concerned, the class `Event` has two responsibilities. One is to indicate the event that occurs, which implies the partial or full-order relationship of the event. The second is to carry the relevant information, i.e. the payload, for the `Callback`s or `Handler`s to process.
As for how to carry information, I prefer to design one or some specific objects for an `Event` because it makes development easier and the code more readable.

---
This time I mainly discuss the `payload` of different events.

After carefully scanning the code logic of the command `train`, I divided its functions into four:
1. training model
2. logging: output the results of the training process, machine status to files and/or terminals
3. tensorboarding: output the model parameters in the training process to Tensorboard processing
4. serializing: save configuration files, training status, model parameters, etc.

Among them, the **training model** is the core responsibility of the command `train`; the latter three are equivalent to three handlers, and their tasks are to deal with some states in the training process in time.

---
Now analyze the data flow of these four functions, namely input and output.
1. training model: **input** has model, data (including datasets and corresponding iterators), optimizer (and its related parameters such as learning rate scheduler); output is trained model, metrics, etc.
2. logging: input in addition to `model`, `data`, `optimizer`, also need some _configuration information__, including what-to-log (what is recorded), how-often-to-log (how often log), when-to-log (when to log once), where-to-log (where the log information is output)
3. tensorboarding: input is similar to `logging`'s input, but the _configuration information_ is slightly less, including: what-to-tensorboard (what should be recorded by tensorboard), when-to-log (when to perform a tensorboard)
4. serializing: input is similar to `logging`'s input, _configuration information_ is slightly different: including what-to-serialize, how-often-to-serialize, when-to-serialize, where (where to save)

It is worth noting that the last three functions have no output, so they will __not__ affect the payload in the `Event`; but also note that the latter three have some __configuration information__ (i.e. internal states of `Handler`s), which requires the `Trainer` to pass to the corresponding object at the right time.

---
Now consider both the event and the data flow, and get the following flow chart (pseudo code):
```python
                        #INPUT  <- args
parse_args()                    #
                        #OUTPUT -> parameter_filename, overrides,     # for constructing params
                        #          serialization_dir, recover, force, # for serializating
                        #          file_friendly_logging              # for logging

                        #INPUT  <- same as the OUTPUT of parse_args
construct_params()              # load a `Params` object from a configuration file (`parameter_filename`) with consideration of `overrides`
                        #OUTPUT -> params,                            # control information
                        #          serialization_dir, recover, force, # for serializating
                        #          file_friendly_logging              # for logging

                        #INPUT  <- same as the OUTPUT of construct_params()
validate_params()               # check the validity of `params`, such as cuda_devices, and the validity of `serialization_dir`, `recover` and `force`
                                #
                                # After,
                                # 1. we set random seeds for reproducible experiments
                                # 2. we create the serialization directory if need
                                # 3. we configure some loggers for logging
                        #OUTPUT -> params,                            # control information
                        #          serializator,                      # for serializating
                        #          loggers,                           # for logging
                        #          tensorboarder                      # for tensorboarding


                        #INPUT <- params,                            # control information
                        #          model, data, optimizer,            # for training
                        #          serializetor,                      # for serializating
                        #          loggers,                           # for logging
                        #          tensorboarder                      # for tensorboarding
                                # Before,
                                # we construct some objects from `params`:
                                # 1. data for training and/or validation, including datasets (what to feed) and iterators (how to feed)
                                # 2. model (what to train)
                                # 3. optimizer (how to train)
                                # (more info)
                                # set some states of the models, such as functions `enable_gradient_clipping` and `no-grad-regexes`
                                #
                                # `serializetor` begins to trace the best metric;
                                # `loggers` print cpu/gpu memory usage and record the time;
                                # `tensorboarder` does nothing
train_model()                   # train models
                        #OUTPUT -> params,                            # control information
                        #          best_model,                        # for evaluating
                        #          metrics,                           # infos
                        #          serializetor,                      # for serializating
                        #          loggers,                           # for logging
                        #          tensorboarder                      # for tensorboarding

                                # Before,
                                # we construct testset and correpsonding data_iterators from `params`.
                        #INPUT  <- params,
                        #          best_model, testset,               # for evaluating
                        #          metrics,                           # infos
                        #          serializetor,                      # for serializating
                        #          loggers,                           # for logging
                        #          tensorboarder                      # for tensorboarding
evaluate_model()                # evaluate the `best_model` on `test_data` and update the `metrics`
                        #OUTPUT -> params,
                        #          metrics,
                        #          serializetor,                      # for serializating
                        #          loggers,                           # for logging
                        #          tensorboarder                      # for tensorboarding
                                # After,
                                # these functions `cleanup_global_logging`, `archive_model` and `dump_metrics` are called.
                                # That is `serilizetor`, `loggers` and `tensorboarder` do some clearing jobs.

                        #INPUT  <- params (including `epoch`)         # control information
                        #          model, data, optimizer,            # for training
                        #          serializetor,                      # for serializating
                        #          loggers,                           # for logging
                        #          tensorboarder                      # for tensorboarding
                                # Before, we record `time`.
[train_model_epoch()]           #
                        #OUTPUT -> epoch_metrics
                                # After, 
                                # `serializetor` dumps this `epoch_metrics`, saves the checkpoint and saves the best_model (if needed);
                                # `loggers` log the peak of memory usage in cpu/gpu, metrics and `training_elapsed_time`;
                                # `tensorboarder` records this `epoch_metrics`;
                                #
                                # `optimizer` is updated (`self._learning_rate_scheduler.step` and `self._momentum_scheduler.step`)

                                # Before,
                                # 1. `loggers` log the peak of memory usage in cpu/gpu;
                                # 2. we prepare some info for `tensorboarder` from `model`;
                                # 3. we get `data_tqdm` for the training batches from `data`;
[[train_model_iteration()]]     #
                        #OUTPUT -> batch_loss
                                # After,
                                # 1. if `batch_loss` is `nan`, raise errors;
                                # 2. `tensorboarder` logs histogram_this_batch and parameter_values;
                                # 3. `serializetor` saves model if needed;
                                # 4. update `optimizer` if needed (`self._learning_rate_scheduler.step_batch`,
                                #                                  `self._momentum_scheduler.step_batch`,
                                #                                  `self._moving_average.apply`);
                                # 5. update the description of `data_tqdm` with the lateset metrics;
```

In order to more clearly reflect the sequence of events, the responsibilities of each object, and the flow of data, I have drawn the following table based on the above pseudo code.
[events.pdf](https://github.com/allenai/allennlp/files/2966351/events.pdf)

The _blue part_ of the table indicates that the Trainer has performed some operations and changed the payload of different events;
The _Comment_ column in the table illustrates the functions of the `Trainer` execution method, the operations performed by the `Handler`s, and the `Callback` function before or after different events.

Hope this form can provide some help.

From the table, we can also observe that `Handler`s often traverse multiple events and need to track some internal states; while `Callback` is for a single event, there is no need to save the internal state, so each `Callback` can be regarded as a stateless `Handler`.

Also note that **not all** of the events are drawn in the table, and some of them will be discussed further below.

---
The `Events` discussed above are caused by the start or completion of a subtask in the process flow. However, it should be noted that there are also two different types of events:
1. Events triggered by _the specified period_;
2. Events due to _exceptions_, such as events caused by `KeyboardInterrupt`.

I call these three classes ProcessEvent, TimerEvent, ExceptionEvent.

The `TimerEvent` is mainly used to periodically save the model during the train_model process. The `Handler`s of the `ExceptionEvent` mainly contain the processing logic of the archive model.
Therefore, the payload of these two types of events is similar to the payload of the event `TRAIN_MODEL_ITERATION_AFTER`.

**Comment by WrRan** [comment]
The issue of `Event` serialization has not been considered in the above discussion, which may affect the backward compatibility of `train --recover`.

**Comment by keitakurita** [comment]
To (hopefully) move the discussion forward, I've created a branch along with a basic implementation of what I envision for the callback system

- [Trainer](https://github.com/keitakurita/allennlp/blob/kkurita-trainer-callbacks/allennlp/training/trainer_with_callbacks.py)
- [Callbacks](https://github.com/keitakurita/allennlp/blob/kkurita-trainer-callbacks/allennlp/training/callbacks.py)

I've been using this system myself the past few weeks and it's working pretty nicely. If the overall idea looks good I'll look into polishing it a bit more and refactoring the Tensorboard and Checkpointer functionality.

**Comment by matt-gardner** [comment]
Just FYI, @joelgrus is traveling for the next couple of days, and I think he's the best person to look at this.  We'll get to it, but it'll be another little while still.

**Comment by kl2806** [comment]
@joelgrus what is the state of this?

**Comment by joelgrus** [comment]
the state of it is that it's a Q2 goal for me that this gets done, but that I need to do some planning/design for it that I haven't done yet

**Comment by keitakurita** [comment]
Hi, I was wondering what the status of this is and if there's been any progress on the design/implementation side. If there's anything I can do to help out, I'd love to jump in but I don't want to duplicate any efforts in progress or go against any design decisions on behalf of the team. Thanks!

**Comment by joelgrus** [comment]
check out #2817

**Comment by matt-gardner** [comment]
Fixed by #2817.
